### DISCUSSION TITLE
ExtensionHost freezes: High CPU usage in `provideTimeline` when switching files - Seeking community insights

### DISCUSSION THREAD

**Discussion by MMMartt** [microsoft/vscode-discussions#2053] [Extension Development QnA] [answered]
Hello fellow VS Code extension developers,

I'm encountering a perplexing issue while developing a VS Code extension and I'm hoping the community might have some insights or similar experiences to share.

Issue Description:
When switching files, particularly in large repositories, the VS Code extension host process occasionally freezes, and CPU usage spikes to 100%. CPU profiles consistently show that `provideTimeline` and `walkTokens` are consuming significant resources.

Key Observations:
1. The issue occurs in the extension host process.
2. It's more likely to happen with my extension enabled, but can occur (less frequently) even when it's disabled. Interestingly, GitLens also seems to increase the likelihood of this issue occurring in my own experience.
3. The problem rarely occurs when all extensions are disabled, but timeline occasionally still freezes.
4. CPU profiling logs are similar whether my extension is enabled or not.
5. More frequent in large repositories or projects with numerous timeline entries.

Reproduction:
- Inconsistent and difficult to reproduce reliably
- More likely when switching files in large repositories
- Occurs both with and without custom extensions, but more frequently with certain extensions active

Environment:
- VS Code version: 1.94.2
- OS: macOS 14.1 (23B2073)
- Hardware: Apple M3 Pro 36 GB

Questions for the Community:
1. Has anyone encountered similar issues or have any insights about potential causes?

2. Do you have any recommendations for effective ways to debug intermittent performance issues like this in VS Code extensions? I'm finding it challenging due to the inconsistent reproduction steps.

Any thoughts, suggestions, or debugging tips would be greatly appreciated!

I've attached a CPU profile for reference. Any insights, experiences, or suggestions would be greatly appreciated. I'm happy to provide more details if needed.
<img width="1085" alt="image" src="https://github.com/user-attachments/assets/b6a92172-ebeb-44f5-a0a2-ab0b40e3d7fa">
[CPU-20241018T070023.086Z.cpuprofile.txt](https://github.com/user-attachments/files/17430288/CPU-20241018T070023.086Z.cpuprofile.txt)

Thank you in advance for your help!

**Comment by MMMartt** [comment]
Update: 

I've just experienced the issue again, this time with all extensions disabled. This leads me to suspect that there might be an underlying problem with the timeline implementation itself. 

However, it's important to note that the frequency of the issue still seems to increase when extensions are enabled.

**Comment by isidorn** [comment, marked as answer]
This sounds like a good vscode issue. Could you open one here please https://github.com/microsoft/vscode/issues
